sui-execution/latest/sui-move-natives/src/eth_state_proof.rs

```diff
@@ -93,17 +114,64 @@ pub fn verify_message_proof(
     let message_map_index = get_message_storage_slot(message.clone(), dwallet_id.clone(), map_slot)
         .map_err(|_| PartialVMError::new(StatusCode::UNKNOWN_STATUS))?;
 
-    let contract_address = Address::from_slice(contract_address.as_slice());
-    let state_root = Bytes32::try_from(state_root.as_slice())
+    let contract_address = String::from_utf8(contract_address.clone())
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let contract_address: Address = contract_address
+        .parse()
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let mut beacon_block: BeaconBlock = serde_json::from_slice(&beacon_block)
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let beacon_block_type = String::from_utf8(beacon_block_type.clone())
         .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
 
+    let beacon_block_type: BeaconBlockType = BeaconBlockType::from_str(&beacon_block_type)
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let mut beacon_block_body =
+        BeaconBlockBodyWrapper::new_from_json(beacon_block_body, &beacon_block_type)
+            .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+    let beacon_block_body: BeaconBlockBody = beacon_block_body.inner();
+
+    let beacon_block_execution_payload =
+        ExecutionPayloadWrapper::new_from_json(beacon_block_execution_payload, &beacon_block_type)
+            .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+    let beacon_block_execution_payload: ExecutionPayload = beacon_block_execution_payload.inner();
+
+    let beacon_block_body = BeaconBlockBody::new_from_existing_with_execution_payload(
+        beacon_block_body,
+        beacon_block_execution_payload,
+    );
+    beacon_block.body = beacon_block_body;
+
+    let eth_state_data = bcs::from_bytes::<ConsensusStateManager<NimbusRpc>>(&eth_state_data)
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let beacon_block_hash = beacon_block
+        .hash_tree_root()
+        .map_err(|_| PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR))?;
+
+    let mut verified_block_header = eth_state_data.get_finalized_header();
+
+    let verified_block_hash = verified_block_header
+        .hash_tree_root()
+        .map_err(|_| PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR))?;
+
+    if beacon_block_hash != verified_block_hash {
+        return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)]));
+    }
+
+    let state_root = beacon_block.body.execution_payload().state_root();
```

Comment:
Please explain the change and why this is required

Reply:
When a user approves a message, he sends the relevant `EthereumState` move object ID, with the whole beacon block that he fetched from an rpc.
In the `EthereumState` object we have an already verified state(otherwise the object wouldn't have been created).
What we do is take the beacon block that is sent from the user, and compute the hash of it, then we want to make sure that the hash is equal to the hash of the header we already have verified in the `EthereumState` object.

If so - then we know we have the correct block, and we can trust it's execution payload input, and then we get the current Execution layer `state_root`.

Reply:
Please document this in the function